### ISSUE TITLE
compound XCAT & YSEQUENCE functions spuriously convert string type to numeric

### ISSUE THREAD

**Issue by karlmsmith** [NOAA-PMEL/Ferret#969]
**Reported by andrew wittenberg on 28 Sep 2009 17:56 UTC**
The following example illustrates spurious conversion of strings to numeric type, when doing concatenation and resequencing in a compound expression.  Strangely, the results are correct when we list the expression directly, but get converted to numeric type when we assign this expression to a variable.

```
        NOAA/PMEL TMAP
        FERRET v6.203  
        Linux(g77) 2.4.21-32 - 06/26/09
        28-Sep-09 13:50     

yes? let a = {"a","b"}
yes? let b = {"x","z"}
yes? let c = xcat(ysequence(a),ysequence(b))
yes? list xcat(ysequence(a),ysequence(b))
             VARIABLE : XCAT(YSEQUENCE(A),YSEQUENCE(B))
             SUBSET   : 2 by 2 points (X-Y)
           1  2  
           1  2
 1   / 1:"a""x"
 2   / 2:"b""z"
yes? list c
             VARIABLE : XCAT(YSEQUENCE(A),YSEQUENCE(B))
             SUBSET   : 2 by 2 points (X-Y)
               1          2      
               1          2
 1   / 1:  1.893E-32  0.000E+00
 2   / 2:  1.893E-32  0.000E+00
```

Defining an intermediate variable appears to make the problem go away:

```
yes? let a = ysequence({"a","b"})
yes? let b = ysequence({"x","z"})
yes? let c = xcat(a,b)
yes? list xcat(a,b)
             VARIABLE : XCAT(A,B)
             SUBSET   : 2 by 2 points (X-Y)
           1  2  
           1  2
 1   / 1:"a""x"
 2   / 2:"b""z"
yes? list c
             VARIABLE : XCAT(A,B)
             SUBSET   : 2 by 2 points (X-Y)
           1  2  
           1  2
 1   / 1:"a""x"
 2   / 2:"b""z"
```

Migrated-From: http://dunkel.pmel.noaa.gov/trac/ferret/ticket/1697

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 28 Sep 2009 20:11 UTC**
Internally, Ferret functions operate on only a single type for a given argument. For the *CAT functions and *SORT functions we've put in a layer of logic which chooses whether to run XCAT or XCAT_STR, etc. so that the user can say either

xcat(astring, bstring), or xcat(vfloat, ufloat).  

In the command-stack operations when Ferret is getting the return type, the type of XCAT is being used rather than XCAT_STR.  

See the call to GCF_RTN_TYPE from MERGE_GRID:

```
 * data type of result
         IF (its_gc) THEN
 * ... can get result data type only when types of all arguments are known
 *     cuz it may be inherited from an argument
           IF (iarg .EQ. 1) THEN
             i = GCF_RTN_TYPE(gcfcn, com_cx)
             IF (i .EQ. pfcn_rtn_string) THEN
               cx_type(mrg_cx) = ptype_string
             ELSE
               cx_type(mrg_cx) = ptype_float
             ENDIF
           ENDIF
```

The fix will be somewhere here. Not fixing for the release of V6.3.

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 30 Sep 2009 18:04 UTC**
More such strangeness, with a different function.  The first LIST appears to do a spurious type conversion, and the last one gives an error.

```
        NOAA/PMEL TMAP
        FERRET v6.203  
        Linux(g77) 2.4.21-32 - 06/26/09
        30-Sep-09 13:57     

yes? let a = {0,0,1,1,1,1,0,1,0}
yes? let b = {"a","b","c","d","e","f","g","h","i"}
yes? list compressi(IF a EQ 0 THEN b)
             VARIABLE : COMPRESSI(IF A EQ 0 THEN B)
             SUBSET   : 9 points (X)
 1   / 1:  1.891E-32
 2   / 2:  0.000E+00
 3   / 3:  1.891E-32
 4   / 4:  0.000E+00
 5   / 5:  1.891E-32
 6   / 6:  0.000E+00
 7   / 7:  1.891E-32
 8   / 8:  0.000E+00
 9   / 9:  1.891E-32
yes? let c = IF a EQ 0 THEN b
yes? list c
             VARIABLE : IF A EQ 0 THEN B
             SUBSET   : 9 points (X)
 1   / 1:"a"
 2   / 2:"b"
 3   / 3:"" 
 4   / 4:"" 
 5   / 5:"" 
 6   / 6:"" 
 7   / 7:"g"
 8   / 8:"" 
 9   / 9:"i"
yes? list compressi(c)
 **ERROR: invalid command: string/numeric data type mix-up: COMPRESSI
```

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 30 Sep 2009 18:35 UTC**
Actually, COMPRESSI  does not operate on a string argument:

```
yes? let h = {"a","b","c",,"e","f",,"h","i"}
 **ERROR: invalid command: string/numeric data type mix-up: COMPRESSI
```

So I'll enter a new ticket requesting STRING versions of COMPRESS functions, including functionality to call these with the same function name as the numeric versions (as done with the *CAT functions).

The same issues we've been seeing at the start of this ticket, with string/float functions will likely still be happening with the string versions of COMPRESS*

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 12 Nov 2010 23:40 UTC**
Just ran into this again, bumping up the priority...

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 8 Dec 2010 23:52 UTC**
As Ansley noted in a 12/3/10 email, calling XCAT_STR directly is a workaround in these compound expressions.  So I am reducing the priority and severity of this bug.

Our remaining interest in resolving this is simply to reduce the number of extraneous function names, in which case sooner would be better than later -- so that we don't build up an inertia of scripts that call XCAT_STR, YCAT_STR, etc. directly.

**Comment by karlmsmith** [comment]
**Comment by steven.c.hankin on 9 Dec 2010 18:50 UTC**
Internally Ferret currently does not have authentic data type sensitivity.  It has only single precision float and string types.  Unless Ansley has put code in for this purpose, the idea of a function that can return a different data type based upon its input type is only "hacked" in.  (Is this wrong?  I don't spot an ef_set_return_type or controls over inheritance of type from args))  Which I think is why it results in bugs.  When full-fledged data type support is added throughout Ferret it will become natural to extend it to strings.

So I ask:  what is the magnitude of down side associated with the user discriminating  XCAT from XCAT_STR?  There will probably be no problem to continue support for XCAT_STR forever -- even after XCAT becomes type sensitive.  So backwards compatibility to old scripts shouldn't become an issue.  Is there a significant use case for removing the XCAT/XCAT_STR distinction based upon how it complicates scripts that can be written today?

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 6 Jul 2017 17:44 UTC**
This is fixed, with the fix to #2547

```

yes? let a = {"a","b"}
yes? let b = {"x","z"}
yes? let c = xcat(ysequence(a),ysequence(b))
yes? list xcat(ysequence(a),ysequence(b))
             VARIABLE : XCAT(YSEQUENCE(A),YSEQUENCE(B))
             SUBSET   : 2 by 2 points (X-Y)
           1  2  
           1  2
 1   / 1:"a""x"
 2   / 2:"b""z"
yes? list c
             VARIABLE : XCAT(YSEQUENCE(A),YSEQUENCE(B))
             SUBSET   : 2 by 2 points (X-Y)
           1  2  
           1  2
 1   / 1:"a""x"
 2   / 2:"b""z"

```

Also, about the side discussion about maintaining XCAT and XCAT_STR: XCAT and the family of *CAT functions now detect the data type and calls XCAT_STR, which is a convenience, but both functions still exist and XCAT_STR can be called.
